opendevin/server/session/agent.py

```diff
@@ -0,0 +1,110 @@
+from typing import Optional
+
+from agenthub.codeact_agent.codeact_agent import CodeActAgent
+from opendevin.controller import AgentController
+from opendevin.controller.agent import Agent
+from opendevin.controller.state.state import State
+from opendevin.core.config import config
+from opendevin.core.logger import opendevin_logger as logger
+from opendevin.core.schema import ConfigType
+from opendevin.events.stream import EventStream
+from opendevin.llm.llm import LLM
+from opendevin.runtime import DockerSSHBox
+from opendevin.runtime.e2b.runtime import E2BRuntime
+from opendevin.runtime.runtime import Runtime
+from opendevin.runtime.server.runtime import ServerRuntime
+
+
+class AgentSession:
+    """Represents a session with an agent.
+
+    Attributes:
+        controller: The AgentController instance for controlling the agent.
+    """
+
+    sid: str
+    event_stream: EventStream
+    controller: Optional[AgentController] = None
+    runtime: Optional[Runtime] = None
+
+    def __init__(self, sid):
+        """Initializes a new instance of the Session class."""
+        self.sid = sid
+        self.event_stream = EventStream(sid)
+
+    async def start(self, start_event: dict):
+        """Starts the agent session.
+
+        Args:
+            start_event: The start event data (optional).
+        """
+        if self.controller or self.runtime:
+            raise Exception(
+                'Session already started. You need to close this session and start a new one.'
+            )
+        await self._create_runtime()
+        await self._create_controller(start_event)
+
+    async def close(self):
+        if self.controller is not None:
+            end_state = self.controller.get_state()
```

Comment:
This is not idempotent. When `close` is called again, `end_state` would be different.

IMHO we should try to make all `close` methods idempotent... seems here we could simply set `self.controller = None` and `self.runtime = None`.

Reply:
Good call. I actually set `_closed=True`, buecause unsetting controller and runtime allows the session to be re-initialized, which could cause problems.